SourceKitLSP: generate Swift textual interfaces for module references

bwhiteley opened the pull request:
When doing a jump to definition on a Swift import statement, generate the textual interface for the module so the IDE can load the generated interface file.

adam-fowler commented:
@ahoppen do you think this will work with my suggested path to displaying symbols in VSCode. https://github.com/swift-server/vscode-swift/issues/408

ahoppen commented on Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift:
```diff
@@ -0,0 +1,39 @@
+//
+//===----------------------------------------------------------------------===//
+
+/// Request a textual interface of a module to display in the IDE.
```
Since this is a non-standard LSP request, could you annotate it with
```swift
/// **(LSP Extension)***
```

ahoppen commented on Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift:
```diff
@@ -0,0 +1,39 @@
+
+  public var uri: DocumentURI
+
+  public init(uri: DocumentURI)
+  {
```
Nitpick to match the style in the rest of the project

```suggestion
  public init(uri: DocumentURI) {
```

ahoppen commented on Sources/SourceKitLSP/Swift/OpenInterface.swift:
```diff
@@ -0,0 +1,85 @@
+    let uri = request.params.textDocument.uri
+    let moduleName = request.params.name
+    self.queue.async {
+      
```
Superfluous newline

ahoppen commented on Sources/SourceKitLSP/Swift/OpenInterface.swift:
```diff
@@ -0,0 +1,85 @@
+  var contents: String
+}
+
+private var textualInterfaces: [String:DocumentURI] = [:]
```
```suggestion
private var textualInterfaces: [String: DocumentURI] = [:]
```

ahoppen commented on Sources/SourceKitLSP/Swift/OpenInterface.swift:
```diff
@@ -0,0 +1,85 @@
+    let moduleName = request.params.name
+    self.queue.async {
+      
+      if let uri = textualInterfaces[moduleName] {
```
I don’t think we can cache textual interfaces like this. The contents of the module might have changed since it was last requested and thus the cache would be out of date. 

I think caching would be safe for system modules that are part of the SDK and are thus very, very unlikely to change (unless you re-install Xcode during a session). IIRC there should be a bit in either the symbol or the module response that says whether the module is a system module.

Also the compiler arguments would need to be part of the key because the generated interface request can return different results for different compiler arguments (think about the generated interface from an Objective-C module that uses `#if FOO` and in one invocation you pass `-DFOO` in the compiler arguments and in the other you don’t.

bwhiteley replied:
I removed caching for now. We could reintroduce it later while addressing the issues you raised.

ahoppen replied:
OK, I think even without caching this is a very good PR.

ahoppen commented on Sources/SourceKitD/SourceKitD.swift:
```diff
@@ -112,6 +112,10 @@ extension SourceKitD {
   }
+  
+  public func cancel(_ handle: sourcekitd_request_handle_t) {
+    api.cancel_request(handle)
```
🎉

ahoppen commented on Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift:
```diff
@@ -0,0 +1,39 @@
+  public static let method: String = "textDocument/openInterface"
+  public typealias Response = InterfaceDetails
+
+  /// The document in which to lookup the module.
```
I think a better description here would be.

```suggestion
  /// The document whose compiler arguments should be used to generate the interface.
```

ahoppen commented on Sources/SourceKitLSP/Swift/OpenInterface.swift:
```diff
@@ -0,0 +1,85 @@
+                              _ uri: DocumentURI,
+                              _ name: String,
+                              _ interfaceURI: DocumentURI,
+                              _ completion: @escaping (Swift.Result<InterfaceInfo, SKDError>) -> Void) {
```
Could you add descriptions for these parameters? Also: Is there any particular reason why you chose to not use argument labels for any of these? I personally always prefer argument label.

ahoppen commented on Sources/SourceKitLSP/Swift/SwiftLanguageServer.swift:
```diff
@@ -106,6 +106,8 @@ public final class SwiftLanguageServer: ToolchainLanguageServer {
 
   let serverOptions: SourceKitServer.Options
+  
+  let generatedInterfacesPath: URL
```
Could you add a doc comment for the property?

ahoppen commented on Sources/SourceKitLSP/Swift/SwiftLanguageServer.swift:
```diff
@@ -259,6 +262,20 @@ public final class SwiftLanguageServer: ToolchainLanguageServer {
+  
+  /// Obtain the path of the directory to write generated module interfaces.
+  private static func generatedInterfacesPath() throws -> URL {
+    if let path = ProcessInfo.processInfo.environment["SOURCEKIT_GENERATED_INTERFACES_PATH"] {
```
What do you think about making this an argument to the `sourcekit-lsp` executable instead of an opaque environment variable?

bwhiteley replied:
I'll make that change.

ahoppen reviewed (commented):
Thanks a lot for the PR. I think the idea of generating the `.swiftinterfaces` into a temporary directory is a good solution for now. @adam-fowler I have a vague memory that we discussed that LSP/VSCode has/will have a notion of generated/transient/non-disk files but I can’t find it anymore. Do you remember anything about it?

The implementation looks good to me overall, I have some comments inline. Also: Could you add a test case for jump to definition to symbols in other modules and a test case for a standalone open interface request?

adam-fowler commented:
> @adam-fowler I have a vague memory that we discussed that LSP/VSCode has/will have a notion of generated/transient/non-disk files but I can’t find it anymore. Do you remember anything about it?

That'll be in the issue I linked to https://github.com/swift-server/vscode-swift/issues/408#issuecomment-1277815702

ahoppen commented:
Is there any precedence from language servers that use these kinds of transient documents or an LSP proposal? I think you are the best person to judge how the contents of the interface should be transferred to the editor. The nice thing about the temporary file implementation is that it will work for all editors.

adam-fowler commented:
> Is there any precedence from language servers that use these kinds of transient documents or an LSP proposal? I think you are the best person to judge how the contents of the interface should be transferred to the editor. The nice thing about the temporary file implementation is that it will work for all editors.

We can go with the temporary file solution. You'll need to be careful to ensure you generate separate versions of these for different swift versions.

bwhiteley commented on Sources/SourceKitLSP/Swift/OpenInterface.swift:
```diff
@@ -0,0 +1,85 @@
+        return
+      }
+      
+      let interfaceFilePath = self.generatedInterfacesPath.appendingPathComponent("\(moduleName).swift")
```
@ahoppen Should these generated interface files have a `.swiftinterface` extension instead of `.swift`?

ahoppen replied:
Oh, yes I think they should.

bwhiteley commented:
Thanks for the feedback. I've addressed the issues with the exception wanting to add a couple of more tests. 

>  add a test case for jump to definition to symbols in other modules

Is there an existing test that is similar to this that you could point me to? I've found a couple of candidates, but one of the challenges I am facing is that some of the tests fail for me even in upstream/main. 
For example:
<img width="1490" alt="image" src="https://user-images.githubusercontent.com/601312/205123498-62875711-3512-4b07-8ca5-ac8369a434a2.png">

ahoppen commented on Sources/SourceKitLSP/Swift/OpenInterface.swift:
```diff
@@ -0,0 +1,84 @@
+    }
+  }
+    
+  /// Open the Swift interface for a module.
+  ///
+  /// - Parameters:
+  ///   - request: The OpenInterfaceRequest.
+  ///   - uri: The document currently being edited.
+  ///   - name: The module name.
+  ///   - interfaceURI:The file where the generated interface should be written.
+  ///   - completion: Completion block to asynchronously receive the InterfaceInfo, or error.
```
I’ve got a couple of minor improvement suggestions for this comment.

```suggestion
  /// Open the Swift interface for a module.
  ///
  /// - Parameters:
  ///   - request: The OpenInterfaceRequest.
  ///   - uri: The document whose compiler arguments should be used to generate the interface.
  ///   - name: The name of the module whose interface should be generated.
  ///   - interfaceURI: The file where the generated interface should be written.
  ///   - completion: Completion block to asynchronously receive the InterfaceInfo, or error.
```

ahoppen commented on Sources/sourcekit-lsp/main.swift:
```diff
@@ -130,6 +130,11 @@ struct Main: ParsableCommand {
   var indexPrefixMappings = [PathPrefixMapping]()
 
+  @Option(
+    name: .customLong("generated-interfaces-path", withSingleDash: true),
```
I would prefer this to have two dashes, like all the options below. `index-store-path` and friends only have a single dash to be compatible with `clang` compiler arguments. And if we do that, I think it’s also not necessary to specify a `name` for the `Option`.

ahoppen commented on Sources/SourceKitLSP/Swift/SwiftLanguageServer.swift:
```diff
@@ -259,6 +263,19 @@ public final class SwiftLanguageServer: ToolchainLanguageServer {
       }
     })
   }
+  
+  /// Obtain the path of the directory to write generated module interfaces.
+  private static func generatedInterfacesPath(overrideInterfacesPath: URL?) throws -> URL {
+    if let overridePath = overrideInterfacesPath {
+      try FileManager.default.createDirectory(at: overridePath, withIntermediateDirectories: true)
+      return overridePath
+    } else {
+      let tempFolderURL = URL(fileURLWithPath: NSTemporaryDirectory())
+      let genFolder = tempFolderURL.appendingPathComponent("GeneratedInterfaces", isDirectory: true)
+      try FileManager.default.createDirectory(at: genFolder, withIntermediateDirectories: true)
+      return genFolder
+    }
+  }
```
What do you think about defaulting the generated interface path to a temporary directory in `sourcekit-lsp/main.swift`. If you do that, the documentation of the `--generated-interfaces-path` could mention that it defaults to a temporary directory, which makes the behavior cleaner IMO.

ahoppen commented:
That’s odd that the tests are failing for you. They are passing fine for me locally. Which Xcode version are you using and are you using a Swift Open Source Toolchain? I’m using Xcode 14.0 with the default toolchain.

I think you can use Tests/SourceKitLSPTests/ImplementationTests.swift as a blueprint for your tests. I don’t think you need to check that many locations but the `impls(at:)` function should give you a good idea of how to invoke an LSP request in the tests.

ahoppen commented on Tests/SourceKitLSPTests/SwiftInterfaceTests.swift:
```diff
@@ -0,0 +1,154 @@
+    let location = try XCTUnwrap(locations.first)
+    XCTAssertTrue(location.uri.pseudoPath.hasSuffix("/Foundation.swiftinterface"))
+    let fileContents = try XCTUnwrap(location.uri.fileURL.flatMap({ try String(contentsOf: $0, encoding: .utf8) }))
+    XCTAssertTrue(fileContents.hasPrefix("import "))
```
Is this testing that `Foundation.swiftinterface` is not empty? I think that’s perfectly fine but a short comment like 
```
// Sanity-check that the generated Swift Interface contains Swift code
```
would explain this, I think.

ahoppen commented on Tests/SourceKitLSPTests/SwiftInterfaceTests.swift:
```diff
@@ -0,0 +1,154 @@
+      try XCTUnwrap(ws.testServer.server?.workspaceForDocument(uri: call.docUri))
+    })
+    let swiftLangServer = try XCTUnwrap(ws.testServer.server?._languageService(for: call.docUri,
+                                                                               .swift,
```
This is being rendered really weirdly on GitHub (the `.` is inside the `s` of `swift`). Is this a normal period or some weird Unicode character?

ahoppen replied:
Hmm, weird. Must have been a rendering issue in Safari then.

ahoppen commented on Tests/SourceKitLSPTests/SwiftInterfaceTests.swift:
```diff
@@ -0,0 +1,154 @@
+    _ = try ws.sk.sendSync(HoverRequest(
+      textDocument: call.docIdentifier,
+      position: Position(line: 0, utf16index: 8)))
+    swiftLangServer.openInterface(request)
```
Is there any reason why this can’t be `ws.sk.sendSync(openInterface)`. If you send the request that way, I think you also don’t need to send the hover request.

bwhiteley replied:
If I do it that way it hits this `fatalError`
https://github.com/apple/sourcekit-lsp/blob/main/Sources/LSPTestSupport/TestJSONRPCConnection.swift#L145

ahoppen replied:
Oh, that’s a very important catch that would have prevented the open interface request from working in the live `sourcekit-lsp` server as well: You need to add `OpenInterfaceRequest.self` to `builtinRequests` in `Messages.swift`.

bwhiteley replied:
I added `OpenInterfaceRequest.self` to `builtinRequests` in `Messages.swift` but still get the same crash.

ahoppen replied:
I just did a little bit more debugging myself and I think these are the changes you need. Essentially, what needs to be done is that SourceKit-LSP has to know about the request type and which method implements the request handling (`registerToolchainTextDocumentRequest`). That also needs to have a fallback response, that's why I changed the return type to be optional.

```diff
diff --git a/Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift b/Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift
index db33f31..6292c1d 100644
--- a/Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift
+++ b/Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift
@@ -14,7 +14,7 @@
 /// **(LSP Extension)**
 public struct OpenInterfaceRequest: TextDocumentRequest, Hashable {
   public static let method: String = "textDocument/openInterface"
-  public typealias Response = InterfaceDetails
+  public typealias Response = InterfaceDetails?
 
   /// The document whose compiler arguments should be used to generate the interface.
   public var textDocument: TextDocumentIdentifier
diff --git a/Sources/SourceKitLSP/SourceKitServer.swift b/Sources/SourceKitLSP/SourceKitServer.swift
index 28f93a7..ef07a3e 100644
--- a/Sources/SourceKitLSP/SourceKitServer.swift
+++ b/Sources/SourceKitLSP/SourceKitServer.swift
@@ -184,6 +184,7 @@ public final class SourceKitServer: LanguageServer {
     registerToolchainTextDocumentRequest(SourceKitServer.completion,
                                          CompletionList(isIncomplete: false, items: []))
     registerToolchainTextDocumentRequest(SourceKitServer.hover, nil)
+    registerToolchainTextDocumentRequest(SourceKitServer.openInterface, nil)
     registerToolchainTextDocumentRequest(SourceKitServer.declaration, .locations([]))
     registerToolchainTextDocumentRequest(SourceKitServer.definition, .locations([]))
     registerToolchainTextDocumentRequest(SourceKitServer.references, [])
@@ -986,6 +987,14 @@ extension SourceKitServer {
     languageService.hover(req)
   }
 
+  func openInterface(
+    _ req: Request<OpenInterfaceRequest>,
+    workspace: Workspace,
+    languageService: ToolchainLanguageServer
+  ) {
+    languageService.openInterface(req)
+  }
+
   /// Find all symbols in the workspace that include a string in their name.
   /// - returns: An array of SymbolOccurrences that match the string.
   func findWorkspaceSymbols(matching: String) -> [SymbolOccurrence] {
@@ -1279,9 +1288,11 @@ extension SourceKitServer {
         let request = Request(openInterface, id: req.id, clientID: ObjectIdentifier(self),
                               cancellation: req.cancellationToken, reply: { (result: Result<OpenInterfaceRequest.Response, ResponseError>) in
           switch result {
-          case .success(let interfaceDetails):
+          case .success(let interfaceDetails?):
             let loc = Location(uri: interfaceDetails.uri, range: Range(Position(line: 0, utf16index: 0)))
             req.reply(.locations([loc]))
+          case .success(nil):
+            req.reply(.failure(.unknown("Could not generate Swift Interface for \(name)")))
           case .failure(let error):
             req.reply(.failure(error))
           }
diff --git a/Tests/SourceKitLSPTests/SwiftInterfaceTests.swift b/Tests/SourceKitLSPTests/SwiftInterfaceTests.swift
index df39e33..44da807 100644
--- a/Tests/SourceKitLSPTests/SwiftInterfaceTests.swift
+++ b/Tests/SourceKitLSPTests/SwiftInterfaceTests.swift
@@ -91,35 +91,19 @@ final class SwiftInterfaceTests: XCTestCase {
       try XCTUnwrap(ws.testServer.server?.workspaceForDocument(uri: importedModule.docUri))
     })
     let swiftLangServer = try XCTUnwrap(ws.testServer.server?._languageService(for: importedModule.docUri, .swift, in: workspace))
-    let expectation = expectation(description: "open interface request")
     let openInterface = OpenInterfaceRequest(textDocument: importedModule.docIdentifier, name: "lib")
-    let request = Request(openInterface, id: .number(1), clientID: ObjectIdentifier(swiftLangServer),
-                          cancellation: CancellationToken(), reply: { (result: Result<OpenInterfaceRequest.Response, ResponseError>) in
-      do {
-        let interfaceDetails = try result.get()
-        XCTAssertTrue(interfaceDetails.uri.pseudoPath.hasSuffix("/lib.swiftinterface"))
-        let fileContents = try XCTUnwrap(interfaceDetails.uri.fileURL.flatMap({ try String(contentsOf: $0, encoding: .utf8) }))
-        XCTAssertTrue(fileContents.contains("""
-          public struct Lib {
-          
-              public func foo()
-          
-              public init()
-          }
-          """))
-      } catch {
-        XCTFail(error.localizedDescription)
-      }
-      expectation.fulfill()
-    })
-    
-    // Send an arbitrary request through the front door first or SourceKitServer won't be properly initialized.
-    _ = try ws.sk.sendSync(HoverRequest(
-      textDocument: importedModule.docIdentifier,
-      position: importedModule.position))
-    swiftLangServer.openInterface(request)
     
-    waitForExpectations(timeout: 15)
+    let interfaceDetails = try XCTUnwrap(ws.sk.sendSync(openInterface))
+    XCTAssertTrue(interfaceDetails.uri.pseudoPath.hasSuffix("/lib.swiftinterface"))
+    let fileContents = try XCTUnwrap(interfaceDetails.uri.fileURL.flatMap({ try String(contentsOf: $0, encoding: .utf8) }))
+    XCTAssertTrue(fileContents.contains("""
+      public struct Lib {
+
+          public func foo()
+
+          public init()
+      }
+      """))
   }
 
   func testSwiftInterfaceAcrossModules() throws {
```

bwhiteley replied:
Thanks for looking into that! I've made the recommended changes.

ahoppen commented on Tests/SourceKitLSPTests/SwiftInterfaceTests.swift:
```diff
@@ -0,0 +1,154 @@
+    }
+    XCTAssertEqual(locations.count, 1)
+    let location = try XCTUnwrap(locations.first)
+    XCTAssertTrue(location.uri.pseudoPath.hasSuffix("/lib.swiftinterface"))
```
Shouldn’t we be jumping to the original source file if it’s in the same package instead of to the Swift interface. AFAICT that’s what we are doing right now. Jumping to the Swift interface here seems like a regression to me.

bwhiteley replied:
It jumps to the Swift interface only if the symbol is a module. Note the `position` in the request is on an import statement. Otherwise it goes to the appropriate line in the source file.

ahoppen replied:
Oh, I didn’t read carefully enough and was assuming that you are jumping to definition on the call. Instead of having the slightly fragile hardcoded position in this test case, the intended design would be to modify `SwiftPMPackage/Sources/exec/main.swift` to

```swift
import /*lib:import*/lib

Lib() . /*Lib.foo:call*/foo()
```

and then you can use define
```diff
- let call = ws.testLoc("Lib.foo:call")
+ let importedModule = ws.testLoc("lib:import")
```
and use `importedModule.position` instead of `Position(line: 0, utf16index: 8)`.

Same for the other test cases in this file.

ahoppen reviewed (commented):
Thanks for adding the test cases. I’ve added a few questions inline.

ahoppen commented on Tests/SourceKitLSPTests/SwiftInterfaceTests.swift:
```diff
@@ -0,0 +1,132 @@
+    
+    let _resp = try sk.sendSync(DefinitionRequest(
+      textDocument: TextDocumentIdentifier(url),
+      position: Position(line: 0, utf16index: 10)))
```
Do you think you could also define a `testLoc` for this instead of hardcoding the position in the tests?

bwhiteley replied:
Unlike the other tests in that file, this particular test doesn't have a TestWorkspace, so `testLoc` is not readily available from what I can see. Rather, the code is provided via `DidOpenTextDocumentNotification`. Similarly, `LocalSwiftTests.swift` is full of `Positions` with hard-coded offsets.

ahoppen replied:
Oh, sorry. I should have read the test case more carefully.

ahoppen reviewed (commented):
Thanks. This looks good to me. Let's get this merged. I've got one final comment inline. Could you also squash your commits after you've addressed the comment to create a nicer git history?

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test

artemcm commented:
Not sure how yet, but I think this is related to a failure in swift-driver Windows CI:
https://github.com/apple/swift-driver/pull/1240
https://ci-external.swift.org/job/swift-driver-PR-windows/227/console 
```
C:\Users\swift-ci\jenkins\workspace\swift-driver-PR-windows\sourcekit-lsp\Sources\LanguageServerProtocol\Messages.swift:59:3: error: cannot find 'OpenInterfaceRequest' in scope
```
